crates/topos/tests/cert_delivery.rs

```diff
@@ -309,3 +313,209 @@ async fn cert_delivery() {
+            .map_err(|_| {
+                vec![format!(
+                    "Unable to submit the certificate to the TCE api on {random_peer}"
+                )]
+            })?;
+
+        tokio::time::sleep(Duration::from_secs(timeout_broadcast)).await;
+        let mut errors = vec![];
+
+        join_all(join_handlers)
+            .await
+            .iter()
+            .for_each(|result| match result {
+                Err(_) => {
+                    errors.push("Unable to properly execute command".to_string());
+                }
+                Ok(Err((peer, error))) => {
+                    errors.push(format!("{peer} {error}"));
+                }
+                _ => {}
+            });
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    })
+    .await
+    .map_err(|error| {
+        warn!("Timeout reached: {:?}", error);
+        error
+    })
+}
+
+async fn run_check_certificate_delivery(
+    number_of_nodes: usize,
+    timeout_broadcast: u64,
+    timeout: u64,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let mut peers_context = create_network(number_of_nodes, vec![]).await;
+
+    let mut status: Vec<bool> = Vec::new();
+
+    for (_peer_id, client) in peers_context.iter_mut() {
+        let response = client
+            .console_grpc_client
+            .status(StatusRequest {})
+            .await
+            .expect("Can't get status");
+
+        status.push(response.into_inner().has_active_sample);
+    }
+
+    assert!(status.iter().all(|s| *s));
+
+    let nodes = peers_context
+        .iter()
+        .map(|peer| peer.1.api_entrypoint.clone())
+        .collect::<Vec<_>>();
+
+    debug!("Nodes used in test: {:?}", nodes);
+
+    let assertion = async move {
+        let peers: Vec<tonic::transport::Uri> = nodes
+            .into_iter()
+            .map(TryInto::try_into)
+            .collect::<Result<_, _>>()
+            .map_err(|e| format!("Unable to parse node list: {e}"))
+            .expect("Valid node list");
+
+        match check_certificate_delivery(timeout_broadcast, peers, timeout).await {
+            Ok(Err(e)) => {
+                panic!("Error with certificate delivery for network of {number_of_nodes}: {e:?}");
+            }
+            Err(e) => {
+                panic!("Timeout elapsed: {e}");
+            }
+            Ok(_) => {
+                info!("Check certificate delivery passed for network of {number_of_nodes}!");
+            }
+        }
+    };
+
+    assertion.await;
+    Ok(())
+}
+
+#[rstest]
+#[test(tokio::test)]
+#[timeout(Duration::from_secs(20))]
+async fn push_and_deliver_cert_5() -> Result<(), Box<dyn std::error::Error>> {
+    run_check_certificate_delivery(5, 5, 20).await
+}
+
+#[rstest]
+#[test(tokio::test)]
+#[timeout(Duration::from_secs(20))]
+async fn push_and_deliver_cert_9() -> Result<(), Box<dyn std::error::Error>> {
+    run_check_certificate_delivery(9, 5, 20).await
+}
```

Comment:
I think you want to use the rstest feature for parameterized test cases with `#[case(args)]` (see [here](https://docs.rs/rstest/latest/rstest/attr.rstest.html#test-parametrized-cases)), easier to read+maintain 

You can see an example in our codebase as well, in the double echo broadcast tests (`crates/topos-tce-broadcast/src/tests/mod.rs`)